### PR TITLE
Pass only retained nodes to MueLu

### LINKED ISSUE #1634: [BUG] MueLu does not work properly with periodic boundary conditions
## Bug Description: What's going wrong?

The system matrix is reduced with the redundant (periodic) dofs, but the coordinates that we pass to muelu is still the full coordinates vector. With trilinos-debug, we run in an internal assertion of Trilinos.

In `tests/input_files/tsi_heatflux_monolithic_pbc.4C.yaml`, the operator has a size of 81 (27*3), but we pass 42 coordinates (instead of `27` or none). The test passes if I skip passing coordinates to muelu.

A potential fix could be to not pass the coordinates to muelu for periodic boundary condition cases. https://github.com/4C-multiphysics/4C/blob/main/src/core/linear_solver/src/preconditioner/4C_linear_solver_preconditioner_teko.cpp#L121

## Interested Parties

@maxfirmbach

### PR DESCRIPTION
**Pull request by amgebauer**
Filtering eliminated nodes before passing them to MueLu. Eliminated nodes occur in periodic boundary conditions.

On the way, I had to fix some `consts` to make it possible.

Closes #1634 

Part of #1600

### CODE REVIEW

File: src/core/fem/src/discretization/4C_fem_discretization_utils.hpp [hunk 1]
```diff
@@ -60,6 +60,17 @@ namespace Core::FE
   std::shared_ptr<Core::LinAlg::MultiVector<double>> extract_node_coordinates(
       const Core::FE::Discretization& discretization, const Core::LinAlg::Map& node_row_map);
 
+  /**
+   * Similar to the other extract_node_coordinates function, but allows to specify a node row map
+   * and filters all eliminated nodes (e.g., in periodic boundary conditions).
```
Comment (mayrmt) [inline src/core/fem/src/discretization/4C_fem_discretization_utils.hpp]:
```suggestion
  /**
   * @brief Compute node map with all retained nodes in case of periodic boundary conditions 
   *
   * Similar to the other extract_node_coordinates function, but allows to specify a node row map
   * and filters all eliminated nodes (e.g., in periodic boundary conditions).
```

**Review by mayrmt — APPROVED** [review]

File: src/core/fem/src/discretization/4C_fem_discretization_utils.hpp (hunk 1 above)
Comment (amgebauer) [reply src/core/fem/src/discretization/4C_fem_discretization_utils.hpp]: Thanks for the suggestion. I reworked the documentation of the method.

**Review by maxfirmbach — APPROVED** [review]
